clvm-derive/src/to_clvm.rs

```diff
@@ -1,53 +1,230 @@
 
     match &ast.data {
-        Data::Struct(data_struct) => match &data_struct.fields {
-            Fields::Named(fields) => {
-                let fields = &fields.named;
-                field_names = fields
-                    .iter()
-                    .map(|field| {
-                        let ident = field.ident.clone().unwrap();
-                        quote!(#ident)
-                    })
-                    .collect();
+        Data::Struct(data_struct) => {
+            if clvm_attr.untagged {
+                panic!("cannot use `untagged` on a struct");
             }
-            Fields::Unnamed(fields) => {
-                let fields = &fields.unnamed;
-                field_names = fields
-                    .iter()
-                    .enumerate()
-                    .map(|(i, _)| {
-                        let index = Index::from(i);
-                        quote!(#index)
-                    })
-                    .collect();
+            let macros = repr_macros(&crate_name, clvm_attr.expect_repr());
+            let field_info = fields(&data_struct.fields);
+            impl_for_struct(&crate_name, &ast, &macros, &field_info)
+        }
+        Data::Enum(data_enum) => {
+            if !clvm_attr.untagged && clvm_attr.repr == Some(Repr::Curry) {
+                panic!("cannot use `curry` on a tagged enum");
+            }
+
+            let mut next_value: Expr = parse_quote!(0);
+            let mut variants = Vec::new();
+
+            for variant in data_enum.variants.iter() {
+                let field_info = fields(&variant.fields);
+                let variant_clvm_attr = parse_clvm_attr(&variant.attrs);
+
+                if variant_clvm_attr.untagged {
+                    panic!("cannot use `untagged` on an enum variant");
+                }
+
+                let repr = variant_clvm_attr
+                    .repr
+                    .unwrap_or_else(|| clvm_attr.expect_repr());
+                if !clvm_attr.untagged && repr == Repr::Curry {
+                    panic!("cannot use `curry` on a tagged enum variant");
+                }
+
+                let macros = repr_macros(&crate_name, repr);
+                let variant_info = VariantInfo {
+                    name: variant.ident.clone(),
+                    value: variant
+                        .discriminant
+                        .as_ref()
+                        .map(|(_, value)| {
+                            if clvm_attr.untagged {
+                                panic!("cannot use `untagged` on an enum with discriminants");
+                            }
+
+                            next_value = parse_quote!(#value + 1);
+                            value.clone()
+                        })
+                        .unwrap_or_else(|| {
+                            let value = next_value.clone();
+                            next_value = parse_quote!(#next_value + 1);
+                            value
```

Comment:
what's going on here? (a comment in the code might be warranted). Presumably if this call fails, it does so before incrementing `next_value`, so it's never done twice. However, it's not obvious in what case we *would* fail here, and why we need to recover from that failure.

Reply:
`variant.discriminant` is an Option since you're free to specify an explicit discriminant on any of the fields. If specified, it's used as the current value. Otherwise, the next incremented value is used as the default, per the Rust enum docs.